### PR TITLE
Fix and document support for dask.distributed

### PR DESCRIPTION
**Pull request by astrofrog**
Prior to this, using dask.distributed did not work, failing with an error about serialization

I've documented the correct way to use dask.distributed (the scheduler has to be explicitly set to the client) and fixed the serialization bugs.

### CODE REVIEW

**Review by keflavich — APPROVED** [review]
Can you comment on some of the changes that aren't directly dask-related?  They look like corrections for other issues.

File: spectral_cube/dask_spectral_cube.py [hunk 1]
```diff
@@ -182,21 +182,24 @@ class FilledArrayHandler:
     """
 
     def __init__(self, cube, fill=np.nan):
-        self._cube = cube
+        self._data = cube._data
+        self._mask = cube._mask
```
Comment (astrofrog) [inline spectral_cube/dask_spectral_cube.py:186]: @keflavich are these the changes you mean? These changes are required for dask.distributed to work - somehow storing a cube on these classes causes serialization issues.

Comment (keflavich) [reply spectral_cube/dask_spectral_cube.py:186]: Yes; got it, thanks, that clarifies it.

**Comment by keflavich** [comment]
I tried this and got spectacular errors:
```python
$ ipython /orange/adamginsburg/ALMA_IMF/reduction/analysis/fullcube_quicklooks.py
Logging to /blue/adamginsburg/adamginsburg/slurmjobs/ipython_log_2021-05-06.py
[TerminalIPythonApp] WARNING | Config option `ignore_old_config` not recognized by `TerminalIPythonApp`.
Activating auto-logging. Current session state plus future input saved.
Filename       : /blue/adamginsburg/adamginsburg/slurmjobs/ipython_log_2021-05-06.py
Mode           : append
Output logging : False
Raw input log  : False
Timestamping   : False
State          : active
/orange/adamginsburg/miniconda3/lib/python3.7/site-packages/IPython/core/magics/logging.py:130: UserWarning: Couldn't start log: Log file is already active: /blue/adamginsburg/adamginsburg/slurmjobs/ipython_log_2021-05-06.py
  warn("Couldn't start log: %s" % sys.exc_info()[1])
 Logging to /blue/adamginsburg/adamginsburg/slurmjobs/ipython_log_2021-05-06.py
Beginning imports
Completed imports
Imported pylab
pbar is <dask.diagnostics.progress.ProgressBar object at 0x2ae272946f28>
Beginning imports
Beginning imports
Beginning imports
Completed imports
Imported pylab
pbar is <dask.diagnostics.progress.ProgressBar object at 0x2b7567d587b8>
Completed imports
Imported pylab
pbar is <dask.diagnostics.progress.ProgressBar object at 0x2b5f9ae59828>
Completed imports
Imported pylab
pbar is <dask.diagnostics.progress.ProgressBar object at 0x2b14c843f7b8>
/orange/adamginsburg/miniconda3/lib/python3.7/site-packages/distributed/node.py:155: UserWarning: Port 8787 is already in use.
Perhaps you already have a cluster running?
Hosting the HTTP server on port 42652 instead
  http_address["port"], self.http_server.port
Task exception was never retrieved
future: <Task finished coro=<_wrap_awaitable() done, defined at /orange/adamginsburg/miniconda3/lib/python3.7/asyncio/tasks.py:596> exception=RuntimeError('\n        An attempt has been made to start a new process before the\n        current process has finished its bootstrapping phase.\n\n        This probably means that you are not using fork to start your\n        child processes and you have forgotten to use the proper idiom\n        in the main module:\n\n            if __name__ == \'__main__\':\n                freeze_support()\n                ...\n\n        The "freeze_support()" line can be omitted if the program\n        is not going to be frozen to produce an executable.')>
Traceback (most recent call last):
  File "/orange/adamginsburg/miniconda3/lib/python3.7/asyncio/tasks.py", line 603, in _wrap_awaitable
    return (yield from awaitable.__await__())
  File "/orange/adamginsburg/miniconda3/lib/python3.7/site-packages/distributed/core.py", line 305, in _
    await self.start()
  File "/orange/adamginsburg/miniconda3/lib/python3.7/site-packages/distributed/nanny.py", line 295, in start
    response = await self.instantiate()
  File "/orange/adamginsburg/miniconda3/lib/python3.7/site-packages/distributed/nanny.py", line 378, in instantiate
    result = await self.process.start()
  File "/orange/adamginsburg/miniconda3/lib/python3.7/site-packages/distributed/nanny.py", line 575, in start
    await self.process.start()
  File "/orange/adamginsburg/miniconda3/lib/python3.7/site-packages/distributed/process.py", line 33, in _call_and_set_future
    res = func(*args, **kwargs)
  File "/orange/adamginsburg/miniconda3/lib/python3.7/site-packages/distributed/process.py", line 203, in _start
    process.start()
  File "/orange/adamginsburg/miniconda3/lib/python3.7/multiprocessing/process.py", line 112, in start
    self._popen = self._Popen(self)
  File "/orange/adamginsburg/miniconda3/lib/python3.7/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/orange/adamginsburg/miniconda3/lib/python3.7/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/orange/adamginsburg/miniconda3/lib/python3.7/multiprocessing/popen_fork.py", line 20, in __init__
    self._launch(process_obj)
  File "/orange/adamginsburg/miniconda3/lib/python3.7/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "/orange/adamginsburg/miniconda3/lib/python3.7/multiprocessing/spawn.py", line 143, in get_preparation_data
    _check_not_importing_main()
  File "/orange/adamginsburg/miniconda3/lib/python3.7/multiprocessing/spawn.py", line 136, in _check_not_importing_main
    is not going to be frozen to produce an executable.''')
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
```

I think this means it doesn't want me spinning up a session from an existing one; it's not clear you can use `LocalCluster` or any distributed client from an already-created cluster

**Comment by astrofrog** [comment]
@keflavich - known limitation of dask.distributed (will find link later) - you need to put the code in your script inside an:

```python
if __name__ == '__main__':
     ...
```

clause. Any luck?

**Pull request merged**